components/shared/code/libs/lib_nvm.c

```diff
@@ -227,30 +242,65 @@ void lib_nvm_init(void)
 
     cycleLog.totalCycles++;
     lib_nvm_requestWrite(NVM_ENTRYID_CYCLE);
+
+#if FEATURE_IS_ENABLED(NVM_TASK)
+    data.queue_handle = xQueueCreateStatic(NVM_ENTRYID_COUNT,
+                                           sizeof(lib_nvm_entryAction_S),
+                                           data.entry_action_queue_buff,
+                                           &data.entry_action_queue);
+#endif
+}
+
+void lib_nvm_start(void)
+{
+#if FEATURE_IS_ENABLED(NVM_TASK)
+    data.rtosStarted = true;
+#endif
+    lib_nvm_check();
 }
 
-void lib_nvm_run(void)
+void lib_nvm_check(void)
 {
     for (lib_nvm_entry_t index = 0U; index < NVM_ENTRYID_COUNT; index++)
     {
-        if (index == NVM_ENTRYID_LOG)
-        {
-            continue;
-        }
-
-        evaluateWriteRequired(index);
-        if (records[index].writeRequired)
+        if (evaluateWriteRequired(index))
         {
-            recordWrite(index);
+            lib_nvm_requestWrite(index);
         }
     }
+}
 
-    evaluateWriteRequired(NVM_ENTRYID_LOG);
-    if (records[NVM_ENTRYID_LOG].writeRequired)
+#if FEATURE_IS_ENABLED(NVM_TASK)
+void lib_nvm_dequeue(void)
+{
+    for (;;)
     {
-        recordWrite(NVM_ENTRYID_LOG);
+        lib_nvm_entryAction_S action;
+
+        if(xQueueReceive(data.queue_handle,
+                         &action,
+                         portMAX_DELAY))
+        {
+            if ((action.entryId == NVM_ENTRYID_LOG) &&
+                uxQueueMessagesWaiting(data.queue_handle))
+            {
+                xQueueSend(data.queue_handle,
+                           (void*)&action,
+                           portMAX_DELAY);
+            }
```

Comment:
what are you doing here? I don't get it. By this point you already know that there's an action queued and you've extracted it into `action` above. `uxQueueMessagesWaiting` returns the number of entries in the queue, so you're just checking if there's still messages left here? And then re-queing the action? What am I missing?

Reply:
Always delay nvm record log writes until it is the last thing to write. It basically ensures we dont write it, write something else, and then have to write it again because the number of writes increased

Reply:
I see, makes sense. You need to document all of this a lot better I think